python/ray/ml/checkpoint.py

```diff
@@ -114,4 +201,392 @@ def to_uri(self, location: str) -> str:
         Returns:
             str: Cloud location containing checkpoint data.
         """
+        path = self.to_directory(path=None)
+        assert is_cloud_target(location)
+        upload_to_bucket(bucket=location, local_path=path)
+        if not isinstance(self, FSStorageCheckpoint):
+            # Only delete directory if it was a temporary directory
+            shutil.rmtree(path)
+        return location
+
+    def to_data_checkpoint(self) -> "DataCheckpoint":
+        raise NotImplementedError
+
+    def to_object_store_checkpoint(self) -> "ObjectStoreCheckpoint":
         raise NotImplementedError
+
+    def to_local_storage_checkpoint(
+        self, path: Optional[str] = None
+    ) -> "LocalStorageCheckpoint":
+        raise NotImplementedError
+
+    def to_external_storage_checkpoint(
+        self, location: str
+    ) -> "ExternalStorageCheckpoint":
+        location = self.to_uri(location=location)
+        return ExternalStorageCheckpoint(location=location)
+
+
+class DataCheckpoint(Checkpoint):
+    def __init__(self, data: Any, metadata: Optional[Dict] = None):
+        Checkpoint.__init__(self, metadata=metadata)
+        self.data = data
+
+    @property
+    def is_fs_checkpoint(self) -> bool:
+        return self.metadata.get("is_fs_checkpoint", False)
+
+    def to_dict(self) -> dict:
+        if isinstance(self.data, dict):
+            return self.data
+        else:
+            return {
+                "data": self.data,
+                "metadata": {**self.metadata, "is_data_checkpoint": True},
+            }
+
+    def to_directory(self, path: Optional[str] = None) -> str:
+        path = path if path is not None else temporary_checkpoint_dir()
+
+        new_metadata = self.metadata.copy()
+
+        os.makedirs(path, exist_ok=True)
+        # Drop marker
+        open(os.path.join(path, ".is_checkpoint"), "a").close()
+        if self.is_fs_checkpoint:
+            # Recover FS checkpoint from data
+            _unpack(self.data["data"], path)
+            new_metadata["is_data_checkpoint"] = False
+        else:
+            # Dump into checkpoint.pkl
+            checkpoint_data_path = os.path.join(path, "checkpoint.pkl")
```

Comment:
Make it a constant? Do we need to document this for our own understanding?

Also does a user need to understand these nuances (where checkpoint.pkl is and is_data_checkpoint v.s. is_fs_checkpoint) to interact with this interface, or this is mostly internal?

Reply:
These are internals, and the `checkpoint.pkl` is just an intermediate representation. Moved to global variables